src/validation/rules/OverlappingFieldsCanBeMerged.js

```diff
@@ -153,34 +185,43 @@ export function OverlappingFieldsCanBeMerged(context: ValidationContext): any {
         visitedFragmentNames,
         subfieldMap
       );
-      const conflicts = findConflicts(subfieldMap);
-      if (conflicts.length > 0) {
-        return [
-          [ responseName, conflicts.map(([ reason ]) => reason) ],
-          conflicts.reduce(
-            (allFields, [ , fields1 ]) => allFields.concat(fields1),
-            [ ast1 ]
-          ),
-          conflicts.reduce(
-            (allFields, [ , , fields2 ]) => allFields.concat(fields2),
-            [ ast2 ]
-          )
-        ];
-      }
+      return subfieldMap;
+    }
+  }
+
+  function subfieldConflicts(
+    conflicts: Array<Conflict>,
+    responseName: string,
+    ast1: Field,
+    ast2: Field
+  ): ?Conflict {
+    if (conflicts.length > 0) {
+      return [
+        [ responseName, conflicts.map(([ reason ]) => reason) ],
+        conflicts.reduce(
+          (allFields, [ , fields1 ]) => allFields.concat(fields1),
+          [ ast1 ]
+        ),
+        conflicts.reduce(
+          (allFields, [ , , fields2 ]) => allFields.concat(fields2),
+          [ ast2 ]
+        )
+      ];
     }
   }
 
   return {
     SelectionSet: {
       // Note: we validate on the reverse traversal so deeper conflicts will be
-      // caught first, for clearer error messages.
+      // caught first, for correct calculation of mutual exclusivity and for
```

Comment:
Yeah. I had the same thought. Maybe I'll add more comments inline near the memoization source to make this more clear. This was already initially here to get the best error messages, and it just happened to simplify memoization.